[#51713] Added storage authentication methods

Kharonus opened the pull request:
If you can read this you are looking for a non-existent, descriptive description of this PR. Sadly, this is still a draft. Please hold the line.


### WAT?

- added one example query auth for every method

Kharonus committed:
[#51713] Added storage authentication methods

- added one example query auth for every method

mereghost commented on modules/storages/app/common/storages/peripherals/storage_interaction/authentication.rb:
```diff
@@ -0,0 +1,140 @@
+          end
+
+          def with_client_credentials(storage:, http_options: {})
+            config = storage.oauth_configuration
+            client_id = config.oauth_client.client_id
+            client_secret = config.oauth_client.client_secret
+            issuer = config.oauth_uri
+            scope = config.scope(user: nil)
+
+            if [client_id, client_secret, issuer, scope].any?(&:blank?)
+              log_message = 'Cannot authenticate storage with client credential oauth flow. Storage not configured.'
+              return error(code: :error, log_message:, payload: storage)
+            end
```
IMO it knows too much about the details of the oauth configuration. WDYT of introducing a `#to_httpx_oauth_config` method (name for sure can be improved) that returns a hash with the required arguments?

Kharonus replied:
like that idea, lemme do something similar

Kharonus replied:
like that idea, lemme do something similar

mereghost commented on modules/storages/app/common/storages/peripherals/storage_interaction/authentication.rb:
```diff
@@ -0,0 +1,140 @@
+
+            current_token = OAuthClientToken.find_by(user_id: user, oauth_client_id: config.oauth_client.id)
+            if current_token.nil?
+              return error(code: :unauthorized,
```
This can be misleading. It isn't unauthorized, it was even tried. :P

Saying that the user isn't authenticated against storage X would be more precise.

Kharonus replied:
actually, that is the only one, that is taken over from previous state and me thinks, it makes sense.

if you never did the login, your are `unauthorized`. It is similar to e.g. how HTTP reacts to the absence of a `Authorization` header.

mereghost commented on modules/storages/app/common/storages/peripherals/oauth_configurations/one_drive_configuration.rb:
```diff
@@ -34,13 +34,14 @@ module OAuthConfigurations
+        attr_reader :oauth_client, :oauth_uri
 
         def initialize(storage)
+          super()
```
Unnecessary call to `super`. There's no initialization there.

Kharonus replied:
Well, if I do not put it here, I need to disable the linter rule. Is it really prettier?

mereghost commented on modules/storages/app/common/storages/peripherals/oauth_configurations/nextcloud_configuration.rb:
```diff
@@ -35,6 +35,7 @@ class NextcloudConfiguration < ConfigurationInterface
         attr_reader :oauth_client
 
         def initialize(storage)
+          super()
```
Unnecessary call to `super`. There's no initialization there.

mereghost commented on modules/storages/app/common/storages/peripherals/oauth_configurations/one_drive_configuration.rb:
```diff
@@ -63,8 +64,8 @@ def extract_origin_user_id(rack_access_token)
-        def scope
-          DEFAULT_SCOPES
+        def scope(user:)
+          user.nil? ? %w[https://graph.microsoft.com/.default] : DEFAULT_SCOPES
```
I think we can always use the `.default` one, as it will grant the permissions allowed by the Azure Application.

It might be worth a test.

mereghost reviewed (commented):
Quick knee-jerk observations.

Kharonus committed:
[#51713] Added new httpx auth implementation

- updated some queries to use new authentication
- restructured tests and VCR cassettes to new auth functionality

Kharonus committed:
[#51713] replaced auth in onedrive queries

Kharonus commented:
closed in favor of a more structured approach in [#51713](https://community.openproject.org/wp/51713).
